unix-ffi/os/os/__init__.py

```diff
@@ -293,24 +298,53 @@ def urandom(n):
     with builtins.open("/dev/urandom", "rb") as f:
         return f.read(n)
 
+class _PopenStream:
+    def __init__(self, fd, pid, mode):
+        self._fd = open(f"/dev/fd/{fd}", mode)
+        self._pid = pid
+        self._closed = False
+        self._exitcode = None
 
-def popen(cmd, mode="r"):
-    import builtins
+    def __enter__(self):
+        return self._fd
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
 
-    i, o = pipe()
-    if mode[0] == "w":
-        i, o = o, i
+    def close(self):
+        if not self._closed:
+            try:
+                self._fd.close()
+            finally:
+                pid, status = waitpid(self._pid, 0)
+                self._exitcode = status
+            self._closed = True
+        return self._exitcode
```

Comment:
This `_PopenStream` object will need to implement read/write/etc operations, so that you can communicate with the process.

Eg prior to this patch it's possible to do:
```py
import os
ls = os.popen("/bin/ls")
print(ls.read())
ls.close()
```
and that still needs to work.